subsys/net/l2/ppp/ppp_l2.c

```diff
@@ -351,22 +350,30 @@ static void net_ppp_mgmt_evt_handler(struct net_mgmt_event_callback *cb, uint32_
 	if (net_if_l2(iface) != &NET_L2_GET_NAME(PPP)) {
 		return;
 	}
+#if defined(CONFIG_NET_PPP)
+	const struct ppp_api *ppp =
+		net_if_get_device(iface)->api;
+#endif
 
 	ctx = net_if_l2_data(iface);
 
-	if (net_if_is_carrier_ok(iface)) {
-		ppp_mgmt_raise_carrier_on_event(iface);
-	} else {
-		ppp_mgmt_raise_carrier_off_event(iface);
-	}
-
 	if (mgmt_event == NET_EVENT_IF_UP) {
+#if defined(CONFIG_NET_PPP)
```

Comment:
Starting and stopping the UART PPP wrapper (ppp.c) is the responsibility of the driver using it, see https://github.com/zephyrproject-rtos/zephyr/blob/5838b62f8914357f164467da7f0f447c20911fe4/drivers/modem/gsm_ppp.c#L629C28-L629C28 PPP L2 communicates in net_pkts, and should remain agnostic to the specific driver wrapping and transporting the net_pkts (in this case ppp.c CONFIG_NET_PPP).

Reply:
But why should there be any extra action needed, to start/stop the ppp.c driver from the user? The ppp.c driver defines a network interface (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/net/ppp.c#L1072), and when the application brings that interface UP, there should be no extra steps needed. That's why I've expressed my concern about removing `enable()` handler in the previous PR (https://github.com/zephyrproject-rtos/zephyr/pull/59124#discussion_r1226628581).

The standard path for all Zephyr network interfaces is:
* Application brings the interface up with `net_if_up()` 
* This triggers `l2_api->enable()` for that particular interface
* L2s `enable()` implementation should take care of any device-specific configuration needed to activate the interface. Just like for Ethernet, we call `start()/stop()` on the underlying Ethernet device, I can't really understand why shouldn't we do the same for PPP.

Right now I'm having the impression we're deviating from standard pattern of dealing with network interfaces, to address some specific, non standard behaviour of the `gsm_ppp.c` module.

Reply:
yes, saw that, it was done in a mentioned PR. Didn't fully understood that change. Was also wondering that net_if_up/down() calls in a driver? Not familiar with gsm_ppp, but maybe it is a driver for modem and more like application/boot up code for ppp net if for accomplishing modem driver tasks. Anyways, this would mean that if application wants to open ppp net if, it needs to start ppp 1st by using device api and then call net_if_up(). I do not feel comfortable with that.

Anyways, I'll let networking stack architects to judge this. My change for this shouldnt break gsm_ppp usage, eventhoug now different than in gsm_ppp usage.

Reply:
Remove the #if defined(CONFIG_NET_PPP) and the start/stop calls are fine.

We are currently using `net_if_carrier_on()/off()` to ensure the PPP L2 subsystem only establishes and drops LCP when the carrier is actually on/off, meaning the everything between the net_if send() API and the host is ready. For a modem, this means its registered to a network, and the device driver is ready to transport PPP frames to said modem.

This is how it works at this time for a modem:

1. Application powers on modem
2. Application brings up L2 network interface (like the one in ppp.c) using `net_if_up()`
3. While the modem is starting up, registering to a network, and having PPP set up, the PPP L2 subsystem does nothing, the carrier is not ready, so trying to start establishing LCP at this time will fail.
4. The modem registers to a network, and prepares PPP
5. The modem invokes `net_if_carrier_on()` to inform the L2 network interface that the carrier is in fact ready to receive net_pkts (wrapped in PPP frames but that is not relevant to the PPP L2 subsystem)
6. Since the carrier is now on, and the interface has been brought up (NET_IF_ADMIN_UP), the NET_IF_UP event is invoked by the net_if subsystem. This is what the PPP L2 subsystem reacts to https://github.com/zephyrproject-rtos/zephyr/blob/cbfa472fa7cfd7b98c8f853b5587337e0bcdc3f1/subsys/net/l2/ppp/ppp_l2.c#L346-L372
7. If the modem looses connection to the network, or is simply powered down, it will call `net_if_carrier_off()`. If this happens, or the application calls `net_if_down()`, the PPP L2 subsystem is informed, and LCP is dropped.

Its important to note that the carrier_on/carrier_off is related to the L2 network interface, not the higher layer LCP, IPCP etc. which relies on the L2 net_if. This seems to be where most of the confusion is coming from, we are not bringing up the L2 subsystem, we are bringing up the PPP data link layer.